### ISSUE TITLE
Issue with wait_for_status_update()?

### ISSUE THREAD

**Issue by otterdahl** [rsmp-nordic/rsmp#3]
While trying to make test cases in rsmp_validator I'm having issues while subscribing to several statuses and using wait_for_status_update(). It never finishes.

I've currently worked around this issue by subscribing, waiting for status update and unsubscribing in sequence for each status, one at a time.

**Comment by emiltin** [comment]
Did you check the log to see if the status update is indeed received? Note that usage like this will wait for a status update with a specific componet, code, name and also value:
```ruby
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
```
Ie. an update with a value that does not exactly match the value of ```status```will be ignored.

**Comment by otterdahl** [comment]
Yes. I've checked the logs.

Right now the function `switch_normal_control` checks S0007, S0011, S0005, one at a time, like this:
```
  # Wait for 'switched on' to be true (dark mode false)
  @status_code_id = 'S0007'
  @status_name = 'status'
  subscribe task,supervisor,site
  status = get_intersection_boolean 'True',intersections
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_dark_mode = get_status_value response
  expect(current_dark_mode).to eq(status)
  unsubscribe_from_all task,supervisor,site

  # Wait for yellow flash status to be false
  @status_code_id = 'S0011'
  @status_name = 'status'
  subscribe task,supervisor,site
  status = get_intersection_boolean 'False',intersections
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_dark_mode = get_status_value response
  expect(current_dark_mode).to eq(status)
  unsubscribe_from_all task,supervisor,site

  # Wait for startup mode to be false
  @status_code_id = 'S0005'
  @status_name = 'status'
  subscribe task,supervisor,site
  status = 'False'
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_controller_starting = get_status_value response
  expect(current_controller_starting).to eq(status)
  unsubscribe_from_all task,supervisor,site
```
It works and produces this:
```
$ be rspec spec/site/command_spec.rb 
Using test config config/swarco.yaml
...

Finished in 1 minute 14.56 seconds (files took 0.27241 seconds to load)
3 examples, 0 failures
```

But If I change it so that it subscribes to all and checks later, it produces error.
```
  # Establish subscriptions
  @status_code_id = 'S0007'
  @status_name = 'status'
  subscribe task,supervisor,site
  @status_code_id = 'S0011'
  @status_name = 'status'
  subscribe task,supervisor,site
  @status_code_id = 'S0005'
  @status_name = 'status'
  subscribe task,supervisor,site

  # Wait for 'switched on' to be true (dark mode false)
  @status_code_id = 'S0007'
  @status_name = 'status'
  status = get_intersection_boolean 'True',intersections
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_dark_mode = get_status_value response
  expect(current_dark_mode).to eq(status)

  # Wait for yellow flash status to be false
  @status_code_id = 'S0011'
  @status_name = 'status'
  status = get_intersection_boolean 'False',intersections
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_dark_mode = get_status_value response
  expect(current_dark_mode).to eq(status)

  # Wait for startup mode to be false
  @status_code_id = 'S0005'
  @status_name = 'status'
  status = 'False'
  response = site.wait_for_status_update component: @component, sCI: @status_code_id, n: @status_name, q:'recent', s:status, timeout: 180
  current_controller_starting = get_status_value response
  expect(current_controller_starting).to eq(status)
  unsubscribe_from_all task,supervisor,site
```
The output is:

```
Using test config config/swarco.yaml
  3m7s    error: #<Async::Task:0x2ab5ed2bd7e8 (failed)> [pid=28252] [2019-12-18 08:55:14 +0100]
               |   Async::TimeoutError: execution expired
               |   → /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/task.rb:64 in `yield'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/condition.rb:38 in `wait'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/probe.rb:50 in `block in capture'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/reactor.rb:266 in `with_timeout'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/probe.rb:49 in `capture'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/archive.rb:50 in `capture'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/site_proxy.rb:194 in `wait_for_status_update'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/site/command_spec.rb:171 in `switch_normal_control'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/site/command_spec.rb:200 in `block (3 levels) in <top (required)>'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/support/test_site.rb:71 in `block in connected'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/support/test_site.rb:16 in `block in within_reactor'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/task.rb:255 in `block in make_fiber'
. 6m17s    error: #<Async::Task:0x2ab5ed6b6278 (failed)> [pid=28252] [2019-12-18 08:58:24 +0100]
               |   Async::TimeoutError: execution expired
               |   → /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/task.rb:64 in `yield'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/condition.rb:38 in `wait'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/probe.rb:50 in `block in capture'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/reactor.rb:266 in `with_timeout'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/probe.rb:49 in `capture'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/archive.rb:50 in `capture'
               |     /var/lib/gems/2.5.0/gems/rsmp-0.1.8/lib/rsmp/site_proxy.rb:194 in `wait_for_status_update'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/site/command_spec.rb:171 in `switch_normal_control'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/site/command_spec.rb:215 in `block (3 levels) in <top (required)>'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/support/test_site.rb:71 in `block in connected'
               |     /home/i0davla/trv/rsmp/rsmp_validator/spec/support/test_site.rb:16 in `block in within_reactor'
               |     /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/task.rb:255 in `block in make_fiber'
..

Finished in 6 minutes 20 seconds (files took 0.27142 seconds to load)
3 examples, 0 failures
```

This is the status update in the log, which should have caused wait_for_status_update() to finish: 
```
2019-12-18 07:52:36 UTC  KK+AG9998     -->  08ab  Received StatusUpdate {"mType":"rSMsg",
"type":"StatusUpdate","mId":"08ab5819-6391-4e7c-9c36-9af44b2df764",
"ntsOId":"KK+AG9998=001TC000",
"xNId":"",
"cId":"KK+AG9998=001TC000",
"sTs":"2019-12-18T07:52:36.000Z",
"sS":[
{"sCI":"S0007","n":"status","s":"True,True","q":"recent"},
{"sCI":"S0005","n":"status","s":"False","q":"recent"},
{"sCI":"S0011","n":"status","s":"False,False","q":"recent"}
]}
```

**Comment by otterdahl** [comment]
Another thing:
Timeout in wait_for_status_update() doesn't seem to cause the test to fail. (it says 0 failures above)

**Comment by otterdahl** [comment]
The TLC groups the subscriptions together in the same StatusUpdate, but only the first sCI is processed in wait_for_status_update()

**Comment by emiltin** [comment]
Fixed in 30d6e8414db3cc. The problem was if a status update contained several statuses it checked all, instead of breaking as soon as a match was found, causes the wait to continue. Will update the gem and the validator.

**Comment by emiltin** [comment]
updated validator. it's now pulling the gem from github, instead of rubygems. this gives a faster turn-around while we develop actively on both them gem and the validator. use 'bundle' as normal to fetch the required gem version.

**Comment by emiltin** [comment]
if the fix works for you, you're welcome to close this issue

**Comment by otterdahl** [comment]
Now it exists prematurely
```
Run options: include {:locations=>{"./spec/site/command_spec.rb"=>[189]}}
F

Failures:

  1) RSMP site commands M0001 set yellow flash
     Failure/Error: expect(current_dark_mode).to eq(status)
     
       expected: "False,False"
            got: "True,True"
     
       (compared using ==)
     # ./spec/site/command_spec.rb:173:in `switch_normal_control'
     # ./spec/site/command_spec.rb:200:in `block (3 levels) in <top (required)>'
     # ./spec/support/test_site.rb:71:in `block in connected'
     # ./spec/support/test_site.rb:16:in `block in within_reactor'
     # /var/lib/gems/2.5.0/gems/async-1.23.0/lib/async/task.rb:255:in `block in make_fiber'

Finished in 3.64 seconds (files took 0.27914 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/site/command_spec.rb:189 # RSMP site commands M0001 set yellow flash
```

**Comment by emiltin** [comment]
Is this method correct? It picks the first item in the sS array, without looking at the code/name:
```
def get_status_value message
  message.attributes['sS'].first['s']
end
```

**Comment by otterdahl** [comment]
You're right, get_status_value is the problem. Fixed in c7455aab30a1937e8b0d21675f4183a6d75a8542 in rsmp_validator.

This issue can be closed

**Comment by emiltin** [comment]
I don't think you have pushed c7455aab to the repo.
